index.html

```diff
@@ -107,27 +125,38 @@ <h2>Methods</h2>
                 <p>Run the following substeps in parallel:</p>
                 <ol>
                   <li>
-                    <p>If <var>sinkId</var> is not the empty string and does not
-                    match any audio output device identified by the result that
-                    would be provided by <a><code>enumerateDevices()</code></a>,
-                    reject <var>p</var> with a new
-                    <a href="https://heycam.github.io/webidl/#idl-DOMException">
-                    <code>DOMException</code></a> whose name is
-                    <a href="https://heycam.github.io/webidl/#notfounderror">
-                    <code>NotFoundError</code></a> and abort these substeps.</p>
+                    <p>Let <var>device</var> be the system default audio output
+                    device.</p>
                   </li>
                   <li>
-                    <p>If <var>sinkId</var> is not the empty string, and the
-                    application would not be permitted to play audio through
-                    the device identified by <var>sinkId</var> if it weren't the
-                    current user agent default device, reject <var>p</var>
-                    with a new <a><code>DOMException</code></a> whose name is
-                    <a href="https://heycam.github.io/webidl/#notallowederror">
-                    <code>NotAllowedError</code></a> and abort these substeps.</p>
-                  </li>
+                    <p>If <var>sinkId</var> is not the empty string, run the
+                    following sub steps:<p>
+                    <ol>
+                      <li>
+                        <p>Set <var>device</var> to the audio output device whose
+                        <code>deviceId</code>, as would be provided by
+                        <a data-cite="GETUSERMEDIA#dom-mediadevices-enumeratedevices">
+                        <code>enumerateDevices()</code></a>, matches <var>sinkId</var>.
+                        If there is no match, reject <var>p</var> with a new
+                        <a data-cite="WEBIDL#idl-DOMException"><code>DOMException</code></a>
+                        whose name is
+                        <a data-cite="WEBIDL#notfounderror"><code>NotFoundError</code></a>,
+                        and abort all sub steps.</p>
+                      </li>
+                      <li>
+                        <p>If <var>document</var> is not <a>permitted to play
+                        audio through device</a> specified by <var>device</var>
+                        and <var>sinkId</var>, then the user agent MUST reject
+                        <var>p</var> with a new
+                        <a data-cite="WEBIDL#idl-DOMException"><code>DOMException</code></a>
+                        whose name is
+                        <a data-cite="WEBIDL#notallowederror"><code>NotAllowedError</code></a>,
+                        and abort these substeps.</p>
```

Comment:
Is this step needed? My mental model is that if a device is exposed by enumerateDevices, it is allowed to be used.
What I thought was missing is how we automatically expose audio output devices in enumerateDevices that are related to an exposed microphone through groupId.

Reply:
> Is this step needed?  ... What I thought was missing is how we automatically expose audio output devices in enumerateDevices that are related to an exposed microphone through groupId.

Note the _**permitted to play audio through device**_ algorithm referenced here is defined on line 200 in this patch, and  _is_ where we deal with microphone through groupId.

> My mental model is that if a device is exposed by enumerateDevices, it is allowed to be used.

Yes, but we cannot rely on JS passing in only those values. JS may attempt to pass in any values it can think of, including previous values it may have stored from earlier sessions, so we should specify that validation algorithm I think.

Also, exactly what is exposed by `enumerateDevices` seems under-specified, if we're going to use that as the defining permission model. (I count 5 mentions of `enumerateDevices` and the only normatively clear statement I find about what it exposes is from the new `selectAudioOutput` API, which says the device the user selected MUST be exposed, which as written doesn't necessarily preclude other devices being listed, especially since it doesn't talk about mics, which is hinted at elsewhere).

Also the permission descriptor tying into the permissions spec gives us two things I think we need:
 1. A way for users to block permission for this site
 2. permissions policy [handling](https://w3c.github.io/permissions/#reading-current-states): *"If there exists a policy-controlled feature"*

That said, I realize the permissions spec leaves a lot up to the UA by default, and am open to language restricting the UA's behavior here.

Reply:
Maybe we should split the PR in two and concentrate on speaker-selection here.
The PR bits related to that seem good to merge to me.

> Note the _**permitted to play audio through device**_ algorithm referenced here is defined on line 200 in this patch, and _is_ where we deal with microphone through groupId.

Right, I think we need that kind of algorithm for getting a proper enumerateDevice list.

> Yes, but we cannot rely on JS passing in only those values. JS may attempt to pass in any values it can think of, including previous values it may have stored from earlier sessions, so we should specify that validation algorithm I think.

The current algorithm is that if the value is not in the enumerateDevices list, we reject.
I am not sure what else we need to cover specifically. 

> Also, exactly what is exposed by `enumerateDevices` seems under-specified

Right, this should complement well the other part of the PR.
It seems we want to more precisely define that:
- enumerateDevices should expose all audio output devices that are already exposed as microphones by enumerateDevices
- enumerateDevices should expose audio output devices exposed by selectAudioOutput.

Reply:
> Maybe we should split the PR in two and concentrate on speaker-selection here.

@youennf good idea. Now done. PTAL.